enhancements/storage/csi-ebs-operator.md

```diff
@@ -0,0 +1,142 @@
+---
+title: csi-ebs-operator
+authors:
+  - "@fbertina"
+reviewers:
+  - "@jsafrane”
+  - "@hekumar"
+  - "@chuffman"
+approvers:
+  - "@..."
+creation-date: 2020-03-09
+last-updated: 2020-03-09
+status: provisional
+see-also:
+replaces:
+superseded-by:
+---
+
+# CSI Operator for AWS EVS
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+In the past storage drivers have been provided either in-tree or via external provisioners. With the move to advocating CSI as the way to deliver storage drivers, Red Hat needs to move the in-tree drivers (and others that our customers need) to use CSI. With the focus on operators in Red Hat OCP 4.x these CSI drivers should be installed in the form of an Operator.
+
+## Motivation
+
+* In-tree drivers will be removed from Kubernetes, we need to continue to make the drivers available and CSI is the way to do this.
+* AWS is a key cloud provider for OpenShift and use of EBS has been supported in both OpenShift 3.x and 4.x, we need to move this driver to be CSI, provided by an operator.
+* CSI drivers (incl. AWS EBS) will enable new storage features, not possible with in-tree volume plugins, like snapshots and volume cloning.
+
+### Goals
+
+* Create an operator to install the AWS EBS CSI driver.
+* Publish the operator on OperatorHub.
+* Package and ship downstream image of AWS EBS CSI driver (upstream: https://github.com/kubernetes-sigs/aws-ebs-csi-driver)
+
+### Non-Goals
+
+* Driver creation (upstream: https://github.com/kubernetes-sigs/aws-ebs-csi-driver)
+
+## Proposal
+
+1. OCP ships a new operator called aws-ebs-csi-driver-operator. This operator:
+   * Installs the AWS EBS CSI driver.
+	 * It creates a new Deployment that runs the controller pieces of the driver (provisioner, attacher, resizer, etc.).
+	 * It creates a new DaemonSet that runs the CSI driver on the worker nodes.
+   * Is installed by OLM once the user opts-in.
+   * Runs in namespace `openshift-aws-ebs-csi-driver-operator`, and its operands run in namespace `openshift-aws-ebs-csi-driver`.
```

Comment:
As I understand OLM, we cannot force namespace name, we can only suggest it. Which makes things more complicated when moving from "optional" to "mandatory" operator.

Reply:
Yeah tested using `olm.targetNamespaces` annotation and it can't be used to enforce namespace selection. One option we have is to exit/throw error when operator is installed in any namespace other than what we recommended.  It could be considered bit heavy handed because at that point operator is already installed and user must undo the work and reinstall the operator in correct namespace. cc @shawn-hurley

Reply:
Could you implement a migration job that would allow for an existing operator to be adopted as the default? 

Is there a use case where you might have to handle this even when mandatory, say I set up a CSI drive for AWS on a bare-metal cluster but then add AWS nodes. Is that a situation that we would need to worry about?

Reply:
> Is there a use case where you might have to handle this even when mandatory, say I set up a CSI drive for AWS on a bare-metal cluster but then add AWS nodes. Is that a situation that we would need to worry about?

Interesting use case. I don't think we need to worry about heterogeneous clusters, at least for now.

But we should worry about "bare metal" clusters on AWS - admin can install OCP on manually prepared AWS virtual machines, without telling the installer they're AWS. We consider such cluster as bare metal. Admin should be allowed to install AWS EBS CSI driver via OLM GUI and provide **their own AWS credentials**, because there is no cloud-credential-operator in the cluster.

So, from OLM perspective, it's just usual optional operator, no changes needed there.

From the operator perspective, we should support manually provided credentials. Perhaps not in the first release.